### ISSUE TITLE
bug report: sometimes deploy verticle future not invoked.

### ISSUE THREAD

**Issue by wujimin** [eclipse/vert.x#2331]
vertx version: 3.5.0  
  
create a verticle:  
```
public void start(Future<Void> startFuture) throw Exception{
  super.start();

  HttpServerOptions options = new HttpServerOptions();
  options.setKeyStoreOptions(new JksOptions().setPath("notExistPath"));
  options.setSsl(true);

  then use options to create a http server
  and httpServer.listen will throw a exception caused by SSLHelper.validate
}
```
  
deploy the verticle:  
```
  vertx.deployVerticle(the verticle class, new DeploymentOptions(), ar->{
    this future will not be invoked.
  });
```
  
in other conditions, eg: port already binded or directly throw exception from verticle.start, deploy future is no problem.

**Comment by vietj** [comment]
with this method you need to complete the `startFuture` by yourself, most likely in the http server listen method callback

**Comment by wujimin** [comment]
you can make your test
even i catch the exception and use startFuture.fail, also can not got deploy future callback.

**Comment by wujimin** [comment]
with this test case:  
1.only "listen catched" invoked  
2."listen callback" and "deploy callback" not invoked  
  
analysis:  
1."SSLHelper.validate" throw exception, in this time, for "HttpServerImpl", "listening" is true and "actualServer" is null  
2.in io.vertx.core.http.impl.HttpServerImpl.close(Handler<AsyncResult<Void>>), "done" will not be invoked cause "listening" and "actualServer" value  
  
```
public class TestVerticle extends AbstractVerticle {
  @Override
  public void start(Future<Void> startFuture) throws Exception {
    super.start();

    HttpServerOptions options = new HttpServerOptions();
    options.setKeyStoreOptions(new JksOptions()
        .setPath("notExistPath"));
    options.setSsl(true);
    try {
      vertx
          .createHttpServer(options)
          .requestHandler(req -> {
          })
          .listen(lh -> {
            System.out.println("listen callback: " + lh.succeeded());
            if (lh.failed()) {
              lh.cause().printStackTrace();
              startFuture.fail(lh.cause());
            }
          });
    } catch (Throwable e) {
      System.out.println("listen catched");
      e.printStackTrace();
      startFuture.fail(e);
    }
  }

  public static void main(String[] args) {
    Vertx vertx = Vertx.vertx();
    vertx.deployVerticle(TestVerticle.class, new DeploymentOptions(), ar -> {
      System.out.println("deploy callback: " + ar.succeeded());
    });
  }
}
```

**Comment by vietj** [comment]
can you provide a reproducer ?

**Comment by wujimin** [comment]
i have provided the reproducer.  
the class TestVerticle  and the main function.  
  
with this test case:  
1.only "listen catched" invoked  
2."listen callback" and "deploy callback" not invoked

**Comment by wujimin** [comment]
hi, anyone can help me to confirm this problem?

**Comment by wujimin** [comment]
can you please reopen this issue first?

**Comment by vietj** [comment]
your reproducer is not valid, it should not call `super.start()` and should complete the future instead

**Comment by wujimin** [comment]
1.infact your super.start() is a empty method, there is no any effect here.  
2.after i delete super.start(), i got the same result.  
3.in my reproducer, can print "listen catched", you can see i had invoke "startFuture.fail(e)"  
  
why not just copy my code and run it?  
just need a little seconds.

**Comment by ruslansennov** [comment]
@wujimin 

> in my reproducer, can print "listen catched", you can see i had invoke "startFuture.fail(e)"

but you don't call `startFuture.complete()`. Try something like this:

```java
            System.out.println("listen callback: " + lh.succeeded());
            if (lh.failed()) {
              lh.cause().printStackTrace();
              startFuture.fail(lh.cause());
            } else {
              startFuture.complete();
            }
```

**Comment by wujimin** [comment]
when someone report this problem to me  
i wrote the reproducer,  traced vertx source code:  
 io.vertx.core.http.impl.HttpServerImpl.close(Handler<AsyncResult>), and found that  "close" ignored all logic because of illegal state  
  
maybe you no need to judge this issue is valid or invalid immediately  
just try it and trace your source code
thanks.

**Comment by wujimin** [comment]
the code never run to listen callback.  
it seems that you did not read my description at all......

**Comment by ruslansennov** [comment]
@vietj @wujimin  I can reproduce this issue. Closing the `HttpServer` in the `catch` block fixes the problem
```java
  @Override
  public void start(Future<Void> startFuture) throws Exception {
    super.start();

    HttpServerOptions options = new HttpServerOptions();
    options.setKeyStoreOptions(new JksOptions()
      .setPath("notExistPath"));
    options.setSsl(true);

    HttpServer server = vertx
      .createHttpServer(options)
      .requestHandler(req -> {
      });
    try {
      server
        .listen(lh -> {
          System.out.println("listen callback: " + lh.succeeded());
          if (lh.failed()) {
            lh.cause().printStackTrace();
            startFuture.fail(lh.cause());
          }
        });
    } catch (Throwable e) {
      System.out.println("listen catched");
      e.printStackTrace();
      server.close();
      startFuture.fail(e);
    }
  }
```

**Comment by wujimin** [comment]
ok, thanks.  
now we can got deploy callback by close httpServer.  
  
and a suggestion：  
it's not so good  
if listen failed, listen callback will not be invoked, and need developers catch exception and close manually  
it still seems to be a bug, but now we have a method to avoid it.

**Comment by sanotosh** [comment]
I got the same issue today (deployment just hangs) and I'm glad I stumbled on this. Instead of `startFuture.fail(e);`, `throw new Exception();` also works. But `server.close();` is key. 

What I found rather inconvenient about this workaround is that I cannot enclose `createHttpServer` in the try block as well. It would be great if someone can submit a PR for this - I'll also do some digging when I have time.
